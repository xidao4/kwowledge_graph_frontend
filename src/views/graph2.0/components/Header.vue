<template>
    <div>
        <a-page-header
                style="background-color: white; border: 1px solid rgb(235, 237, 240)"
                title="编辑图谱"
                sub-title=""
                @back="back"
        >
            <template slot="extra">
                <a-button key="3" type="primary" style="margin-right: 10px;" :loading="loading" @click="save">
                    {{btnText}}
                </a-button>
                <a-dropdown>
                    <a-menu slot="overlay" @click="handleTypeChange">
                        <a-menu-item :key="item.value" v-for="item in downloadFileType">
                            <a-icon :type="item.icon" />
                            {{item.type}}
                        </a-menu-item>
                    </a-menu>
                    <a-button type="primary" style="margin-left: 8px">
                        下载 <a-icon type="down" /> </a-button>
                </a-dropdown>
            </template>
            <template slot="footer">
                <a-tabs :default-active-key="currentGraphId" @change="changeCurrentGraph">
                    <a-tab-pane :key="graphIds.typesetting" tab="排版模式" />
                    <a-tab-pane :key="graphIds.force" tab="力引导模式" />
                </a-tabs>
            </template>
            <DownloadImgModal :type="chosenFileType"></DownloadImgModal>
            <DownloadFileModal :type="chosenFileType"></DownloadFileModal>
        </a-page-header>
        <Handler
                v-if="false"
                class="handler"
                mode="horizontal"
                position="bottom"
                :expand="isExpand"
                :callback="toggleHandler"
        >
        </Handler>
    </div>
</template>

<script>
    import { mapGetters, mapActions, mapMutations } from 'vuex';
    import { message } from 'ant-design-vue'
    import DownloadImgModal from "./downloadModal/DownloadImgModal";
    import DownloadFileModal from "./downloadModal/DownloadFileModal";
    import Handler from "./Handler";
    const fileType = {
        png: 'png',
        jpeg: 'jpeg',
        xml: 'xml',
        rdf: 'rdf',
        owl: 'owl',
        json: 'json',
        webp: 'webp',
        bmp: 'bmp'
    };
    const btnTextType = {
        save: '保存',
        saveLoading: '保存中...',
    };
    export default {
        name: "Header",
        components: {
            Handler,
            DownloadFileModal,
            DownloadImgModal,
        },
        data(){
            return {
                downloadFileType: [
                    {
                        type: 'PNG图片（*.png）',
                        value: fileType.png,
                        icon: 'file-image'
                    },
                    {
                        type: 'JPEG图片（*.jpeg）',
                        value: fileType.jpeg,
                        icon: 'file-image'
                    },
                    {
                        type: 'WEBP图片（*.webp）',
                        value: fileType.webp,
                        icon: 'file-image'
                    },
                    {
                        type: 'BMP图片（*.bmp）',
                        value: fileType.bmp,
                        icon: 'file-image'
                    },
                    {
                        type: 'JSON文件（*.json）',
                        value: fileType.json,
                        icon: 'file-text'
                    },
                ],
                chosenFileType: '',
                loading: false,
                btnText: btnTextType.save,
                showDownloadImg: false,
                showDownloadJson: false,
                isExpand: true,
            }
        },
        computed: {
            ...mapGetters([
                'graphIds',
                'currentGraphId',
                'forceGraph',
                'typesettingGraph',
                'currentGraph',
                'showDownloadImgModal',
                'showDownloadJsonModal',
            ])
        },
        methods: {
            ...mapMutations([
                'set_currentGraphId',
                'set_currentGraph',
                'set_showDownloadImgModal',
                'set_showDownloadFileModal',
            ]),
            ...mapActions({
                saveAct: 'save',
                thumbnail: 'thumbnail'
            }),
            handleTypeChange(value) {
                this.chosenFileType = value.key;
                if(null === this.currentGraph) {
                    message.error("图谱不存在，请重新加载！")
                } else {
                    if (this.chosenFileType === fileType.png || this.chosenFileType === fileType.jpeg
                        || this.chosenFileType === fileType.webp || this.chosenFileType === fileType.bmp) {
                        this.set_showDownloadImgModal(true);
                    } else if (this.chosenFileType === fileType.json) {
                        this.set_showDownloadFileModal(true);
                    }
                }
            },
            changeCurrentGraph(key){
                this.set_currentGraphId(key);
                if(key === this.graphIds.force){
                    // this.set_currentGraph(this.forceGraph)
                } else if(key === this.graphIds.typesetting){
                    this.set_currentGraph(this.typesettingGraph)
                }
            },
            toggleHandler (val) {
                const _t = this
                _t.isExpand = val !== undefined ? val : !_t.isExpand
            },
            async save(){
                this.loading = true;
                this.btnText = btnTextType.saveLoading;
                await this.saveAct();
                this.loading = false;
                this.btnText = btnTextType.save;
            },
            back(){
                let imgUrl = this.currentGraph.toDataURL('image/jpeg', '#FFFFFF');
                this.$router.go(-1);
                this.thumbnail(imgUrl);
            }
        },
    }
</script>

<style scoped>
tr:last-child td {
    padding-bottom: 0;
}
#components-page-header-demo-responsive .content {
    display: flex;
}
#components-page-header-demo-responsive .ant-statistic-content {
    font-size: 20px;
    line-height: 28px;
}
@media (max-width: 576px) {
    #components-page-header-demo-responsive .content {
        display: block;
    }

    #components-page-header-demo-responsive .main {
        width: 100%;
        margin-bottom: 12px;
    }

    #components-page-header-demo-responsive .extra {
        width: 100%;
        margin-left: 0;
        text-align: left;
    }
}
</style>
