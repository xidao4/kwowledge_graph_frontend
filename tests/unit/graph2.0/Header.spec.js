import { mount, createLocalVue } from '@vue/test-utils'
import Header from '@/views/graph2.0/components/Header.vue'
import Antd from 'ant-design-vue';
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(Antd);
localVue.use(Vuex);

describe('Header.vue', () => {
    let getters;
    let mutations;
    let actions;
    let store;
    let wrapper;
    const $router = {
        go: jest.fn()
    };

    beforeEach(() => {
        getters = {
            showDownloadImgModal: () => true,
            showDownloadFileModal: () => true,
            currentGraphId: () => 'typesetting',
            graphIds: () => {
                return {
                    force: 'force',
                    typesetting: 'typesetting'
                }
            },
            currentGraph: () => {
                return {
                    data: jest.fn(),
                    updateLayout: jest.fn(),
                    toDataURL: jest.fn()
                }
            },
            forceGraph: () => {
                return {
                    data: jest.fn(),
                    updateLayout: jest.fn(() => {
                        console.log('force update')
                    }),
                    render: jest.fn()
                }
            },
            typesettingGraph: () => {
                return {
                    data: jest.fn(),
                    updateLayout: jest.fn(() => {
                        console.log('typesetting update')
                    }),
                    render: jest.fn()
                }
            },
        };
        mutations = {
            set_currentGraphId: jest.fn(),
            set_currentGraph: jest.fn(),
            set_showDownloadImgModal: jest.fn(),
            set_showDownloadFileModal: jest.fn(),
            clear_graphs: jest.fn(),
            set_isNew: jest.fn()
        };
        actions = {
            save: jest.fn(),
            thumbnail: jest.fn()
        };
        store = new Vuex.Store({
            state: {},
            getters,
            mutations,
            actions
        });
        wrapper = mount(Header, {
            store,
            localVue,
            mocks: {
                $router
            }
        });
    });
    it('header snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('保存',  () => {
        global.URL.revokeObjectURL = jest.fn();
        let buttons = wrapper.findAll('button');
        for(let i = 0; i < buttons.length; i++){
            let button = buttons.at(i);
            button.trigger('click');
        }
        expect(actions.save).toHaveBeenCalled();
    });
    it('返回',  () => {
        let button = wrapper.find('.ant-page-header-back-button');
        button.trigger('click');
        expect(actions.thumbnail).toHaveBeenCalled();
    });
    it('模式切换',  () => {
        let tabs = wrapper.findAll('.ant-tabs-tab');
        for(let i = 0; i < tabs.length; i++){
            let tab = tabs.at(i);
            tab.trigger('click');
        }
        expect(mutations.set_currentGraphId).toHaveBeenCalled();
    });
});

