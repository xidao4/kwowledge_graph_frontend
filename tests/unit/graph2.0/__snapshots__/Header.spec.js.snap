// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.vue header snapshot 1`] = `
<div>
  <div class="ant-page-header has-footer ant-page-header-ghost" style="background-color: white; border: 1px solid rgb(235, 237, 240);">
    <div class="ant-page-header-heading">
      <div class="ant-page-header-back">
        <div role="button" tabindex="0" style="border: 0px; background: transparent; padding: 0px; line-height: inherit; display: inline-block;" aria-label="Back" class="ant-page-header-back-button"><i aria-label="icon: arrow-left" class="anticon anticon-arrow-left"><svg viewBox="64 64 896 896" data-icon="arrow-left" width="1em" height="1em" fill="currentColor" aria-hidden="true" focusable="false" class="">
              <path d="M872 474H286.9l350.2-304c5.6-4.9 2.2-14-5.2-14h-88.5c-3.9 0-7.6 1.4-10.5 3.9L155 487.8a31.96 31.96 0 0 0 0 48.3L535.1 866c1.5 1.3 3.3 2 5.2 2h91.5c7.4 0 10.8-9.2 5.2-14L286.9 550H872c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"></path>
            </svg></i></div>
      </div><span class="ant-page-header-heading-title">编辑图谱</span><span class="ant-page-header-heading-extra"><button type="button" class="ant-btn ant-btn-primary" style="margin-right: 10px;"><span>保 存</span></button> <button type="button" class="ant-btn ant-btn-primary ant-dropdown-trigger" style="margin-left: 8px;"><span>下载</span><i aria-label="icon: down" class="anticon anticon-down"><svg viewBox="64 64 896 896" data-icon="down" width="1em" height="1em" fill="currentColor" aria-hidden="true" focusable="false" class="">
            <path d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"></path>
          </svg></i></button></span>
    </div>
    <div class="ant-page-header-content"> </div>
    <div class="ant-page-header-footer">
      <div class="ant-tabs ant-tabs-top ant-tabs-line">
        <div role="tablist" tabindex="0" class="ant-tabs-bar ant-tabs-top-bar">
          <div class="ant-tabs-nav-container"><span unselectable="unselectable" class="ant-tabs-tab-prev ant-tabs-tab-btn-disabled"><span class="ant-tabs-tab-prev-icon"><i aria-label="icon: left" class="anticon anticon-left ant-tabs-tab-prev-icon-target"><svg viewBox="64 64 896 896" data-icon="left" width="1em" height="1em" fill="currentColor" aria-hidden="true" focusable="false" class=""><path d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 0 0 0 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"></path></svg></i></span></span><span unselectable="unselectable" class="ant-tabs-tab-next ant-tabs-tab-btn-disabled"><span class="ant-tabs-tab-next-icon"><i aria-label="icon: right" class="anticon anticon-right ant-tabs-tab-next-icon-target"><svg viewBox="64 64 896 896" data-icon="right" width="1em" height="1em" fill="currentColor" aria-hidden="true" focusable="false" class=""><path d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"></path></svg></i></span></span>
            <div class="ant-tabs-nav-wrap">
              <div class="ant-tabs-nav-scroll">
                <div class="ant-tabs-nav ant-tabs-nav-animated">
                  <div>
                    <div role="tab" aria-disabled="false" aria-selected="true" class="ant-tabs-tab-active ant-tabs-tab">排版模式</div>
                    <div role="tab" aria-disabled="false" aria-selected="false" class=" ant-tabs-tab">力引导模式</div>
                  </div>
                  <div class="ant-tabs-ink-bar ant-tabs-ink-bar-animated" style="display: block; transform: translate3d(0px,0,0); webkit-transform: translate3d(0px,0,0); width: 0px;"></div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="presentation" style="width: 0px; height: 0px; overflow: hidden; position: absolute;"></div>
        <div class="ant-tabs-content ant-tabs-content-animated ant-tabs-top-content" style="margin-left: 0%;">
          <div role="tabpanel" aria-hidden="false" class="ant-tabs-tabpane ant-tabs-tabpane-active">
            <div tabindex="0" role="presentation" style="width: 0px; height: 0px; overflow: hidden; position: absolute;"></div>
            <div tabindex="0" role="presentation" style="width: 0px; height: 0px; overflow: hidden; position: absolute;"></div>
          </div>
          <div role="tabpanel" aria-hidden="true" class="ant-tabs-tabpane ant-tabs-tabpane-inactive"></div>
        </div>
        <div tabindex="0" role="presentation" style="width: 0px; height: 0px; overflow: hidden; position: absolute;"></div>
      </div>
    </div>
  </div>
</div>
`;
